[JIT] Support multiple outputs in subgraph matcher.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48992 [JIT] Support multiple outputs in subgraph matcher.**



Differential Revision: [D25388100](https://our.internmc.facebook.com/intern/diff/D25388100)

Commit:
[WIP][JIT] Support multiple outputs in subgraph matcher.

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Support multiple outputs in subgraph matcher."


Differential Revision: [D25388100](https://our.internmc.facebook.com/intern/diff/D25388100)

[ghstack-poisoned]

Commit:
Update on "[JIT] Support multiple outputs in subgraph matcher."




Differential Revision: [D25388100](https://our.internmc.facebook.com/intern/diff/D25388100)

[ghstack-poisoned]

Commit:
Update on "[JIT] Support multiple outputs in subgraph matcher."




Differential Revision: [D25388100](https://our.internmc.facebook.com/intern/diff/D25388100)

[ghstack-poisoned]

bwasti commented on test/cpp/jit/test_subgraph_matcher.cpp:
```diff
@@ -507,8 +509,57 @@ graph(%x):
   %z = my::op2(%x)
   return (%y, %z))IR",
       &pattern2);
+  // Not supported multi-output pattern
```
why isn't this supported?

ZolotukhinM replied:
Because not the whole pattern is covered by a traversal up from the first output (`%z = ...` is not visited). I will add a comment.

bwasti commented on torch/csrc/jit/ir/subgraph_matcher.cpp:
```diff
@@ -59,16 +62,26 @@ bool patternGraphIsValid(const Graph& pattern) {
+}
+
+bool SubgraphMatcher::isPatternOutput(const Value* v) const {
+  if (v->owningGraph() != &pattern_) {
```
I don't think it will hurt, but is this check necessary?

bwasti commented on torch/csrc/jit/ir/subgraph_matcher.h:
```diff
@@ -45,6 +45,27 @@ struct Match {
  * but the latter is taken as non-const so that Match may contain non-const
  * pointers.  This enables clients of this API to use Match to drive mutations.
+ *
+ * Note [Multi-output Patterns]
```
the non-backtracking algorithm (`O(M*N)` I think: BFS at each node in the graph) or endowing an output ordering (via some kind of topo-sort) might be a good bootcamp task :)

ZolotukhinM replied:
Mm, I don't know if such an algorithm exists, maybe I'm missing something? Will be happy to discuss it :)

bwasti commented on torch/csrc/jit/ir/subgraph_matcher.h:
```diff
@@ -45,6 +45,27 @@ struct Match {
+ * output and assume that it is sufficient to traverse up from it to match the
+ * entire pattern.
+ *
+ * Corrolary 1: the order of outputs in the pattern matters!
```
does the order in which internal outputs are consumed matter?

```
graph(a, b):
  c = node0(a)
  d = node1(b)
  return c + d, d
```
vs
```
graph(a, b):
  d = node1(b)
  c = node0(a)
  return c + d, d
```

ZolotukhinM replied:
No, that does not matter.

bwasti reviewed (approved):
lgtm

Commit:
Update on "[JIT] Support multiple outputs in subgraph matcher."




Differential Revision: [D25388100](https://our.internmc.facebook.com/intern/diff/D25388100)

[ghstack-poisoned]

Commit:
Update on "[JIT] Support multiple outputs in subgraph matcher."




Differential Revision: [D25388100](https://our.internmc.facebook.com/intern/diff/D25388100)

[ghstack-poisoned]

Commit:
Update on "[JIT] Support multiple outputs in subgraph matcher."




Differential Revision: [D25388100](https://our.internmc.facebook.com/intern/diff/D25388100)

[ghstack-poisoned]
